sql/query/expr.go

```diff
@@ -0,0 +1,651 @@
+				return uintOp, null, null, err
+			}
+			right, err = right.ToUint()
+			if err != nil {
+				return uintOp, null, null, err
+			}
+			return uintOp, left, right, nil
+		}
+		if left.IntVal != nil && right.IntVal != nil {
+			return intOp, left, right, nil
+		}
+	}
+
+	left, err = left.ToFloat()
+	if err != nil {
+		return floatOp, null, null, err
+	}
+	right, err = right.ToFloat()
+	if err != nil {
+		return floatOp, null, null, err
+	}
+	return floatOp, left, right, nil
+}
+
+func evalBinaryExpr(expr *parser.BinaryExpr, env Env) (sqlwire.Datum, error) {
+	left, err := EvalExpr(expr.Left, env)
+	if err != nil {
+		return null, err
+	}
+	right, err := EvalExpr(expr.Right, env)
+	if err != nil {
+		return null, err
+	}
+
+	// TODO(pmattis): Overflow/underflow checks?
+
+	var typ opType
+
+	switch expr.Operator {
+	case parser.Bitand:
+		typ, left, right, err = prepareBinaryArgs(intOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal &= *right.UintVal
+		case intOp:
+			*left.IntVal &= *right.IntVal
+		}
+		return left, nil
+
+	case parser.Bitor:
+		typ, left, right, err = prepareBinaryArgs(intOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal |= *right.UintVal
+		case intOp:
+			*left.IntVal |= *right.IntVal
+		}
+		return left, nil
+
+	case parser.Bitxor:
+		typ, left, right, err = prepareBinaryArgs(intOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal ^= *right.UintVal
+		case intOp:
+			*left.IntVal ^= *right.IntVal
+		}
+		return left, nil
+
+	case parser.Plus:
+		typ, left, right, err = prepareBinaryArgs(floatOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal += *right.UintVal
+		case intOp:
+			*left.IntVal += *right.IntVal
+		case floatOp:
+			*left.FloatVal += *right.FloatVal
+		}
+		return left, nil
+
+	case parser.Minus:
+		typ, left, right, err = prepareBinaryArgs(floatOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			// If the unsigned subtraction would result in a negative number, convert
+			// to signed values.
+			if *right.UintVal > *left.UintVal {
+				v := -int64(*right.UintVal - *left.UintVal)
+				left = sqlwire.Datum{IntVal: &v}
+			} else {
+				*left.UintVal -= *right.UintVal
+			}
+		case intOp:
+			*left.IntVal -= *right.IntVal
+		case floatOp:
+			*left.FloatVal -= *right.FloatVal
+		}
+		return left, nil
+
+	case parser.Mult:
+		typ, left, right, err = prepareBinaryArgs(floatOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal *= *right.UintVal
+		case intOp:
+			*left.IntVal *= *right.IntVal
+		case floatOp:
+			*left.FloatVal *= *right.FloatVal
+		}
+		return left, nil
+
+	case parser.Div:
+		// Division always operates on floats. TODO(pmattis): Is this correct?
```

Comment:
Postgres and Microsoft return an int when you divide two ints. MySQL converts to decimal and has a separate `DIV` operator for integer division. For once, I think MySQL gets it right.